pkg/k8s/rule_translate.go

```diff
@@ -0,0 +1,180 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/cilium/cilium/common/types"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/policy/api"
+)
+
+var _ policy.Translator = RuleTranslator{}
+
+// RuleTranslator implements pkg/policy.Translator interface
+// Translate populates/depopulates given rule with ToCIDR rules
+// Based on provided service/endpoint
+type RuleTranslator struct {
+	Service  types.K8sServiceNamespace
+	Endpoint types.K8sServiceEndpoint
+	Revert   bool
+}
+
+// Translate calls TranslateEgress on all r.Egress rules
+func (k RuleTranslator) Translate(r *api.Rule) error {
+	for egressIndex := range r.Egress {
+		err := k.TranslateEgress(&r.Egress[egressIndex])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// TranslateEgress populates/depopulates egress rules with ToCIDR entries based
+// on toService entries
+func (k RuleTranslator) TranslateEgress(r *api.EgressRule) error {
+	err := k.depopulateEgress(r)
+	if err != nil {
+		return err
+	}
+	if !k.Revert {
+		err := k.populateEgress(r)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (k RuleTranslator) populateEgress(r *api.EgressRule) error {
+	for _, service := range r.ToServices {
+		// TODO: match services by labels
+		if service.K8sService == api.K8sServiceNamespace(k.Service) {
+			if err := generateToCidrFromEndpoint(r, k.Endpoint); err != nil {
+				return err
+			}
+			// TODO: generateToPortsFromEndpoint when ToPorts and ToCIDR are compatible
+		}
+	}
+	return nil
+}
+
+func (k RuleTranslator) depopulateEgress(r *api.EgressRule) error {
+	for _, service := range r.ToServices {
+		// TODO: match services by labels
+		if service.K8sService == api.K8sServiceNamespace(k.Service) {
+			if err := deleteToCidrFromEndpoint(r, k.Endpoint); err != nil {
+				return err
+			}
+			// TODO: generateToPortsFromEndpoint when ToPorts and ToCIDR are compatible
+		}
+	}
+	return nil
+}
+
+// generateToCidrFromEndpoint takes an egress rule and populates it with
+// ToCIDR rules based on provided endpoint object
+func generateToCidrFromEndpoint(
+	egress *api.EgressRule, endpoint types.K8sServiceEndpoint) error {
+
+	// This will generate one-address CIDRs consisting of endpoint backend ip
+	mask := net.CIDRMask(128, 128)
+	for ip := range endpoint.BEIPs {
+		epIP := net.ParseIP(ip)
+		if epIP == nil {
+			return fmt.Errorf("Unable to parse ip: %s", ip)
+		}
+
+		found := false
```

Comment:
I wouldn't care too much about de-duplicating CIDR IP address prefixes. @ianvernon has a change ready to de-duplicate and aggregate all CIDRs afterwards.

Reply:
I saw his PR after implementing that, will definitely use it.

Reply:
His PR is not yet merged in though, but I will ad an issue to track adding the use of his code here.

Reply:
OK, LGTM.